siliconcompiler/tools/yosys/syn_fpga.tcl

```diff
@@ -1,10 +1,124 @@
+                        -cell \$_DFFE_PN?P_ 01 \
+                        -cell \$_DFFSR_PNN_ 01 \
+                        -cell \$_DFFSRE_PNNP_ 01 \
+
+                } else {
+                    yosys log "Legalize DFFs for flop enable"
+                    yosys log "Legalize DFFs for async set"
+                    yosys dfflegalize \
+                        -cell \$_DFF_P_ 01 \
+                        -cell \$_DFF_PN1_ 01 \
+                        -cell \$_DFFE_PP_ 01 \
+                        -cell \$_DFFE_PN1P_ 01 \
+
+                }
+            } else {
+                if { $async_reset == 1 } {
+                    yosys log "Legalize DFFs for flop enable"
+                    yosys log "Legalize DFFs for async reset"
+                    yosys dfflegalize \
+                        -cell \$_DFF_P_ 01 \
+                        -cell \$_DFF_PN0_ 01 \
+                        -cell \$_DFFE_PP_ 01 \
+                        -cell \$_DFFE_PN0P_ 01 \
+
+                } else {
+                    # Choose to legalize to async resets even though they won't
+                    # tech map to get the user to fix their code and put in
+                    # synchronous resets
+                    yosys log "Legalize DFFs for flop enable"
+                    yosys dfflegalize \
+                        -cell \$_DFF_P_ 01  \
+                        -cell \$_DFF_P??_ 01 \
+                        -cell \$_DFFE_PP_ 01 \
+                        -cell \$_DFFE_P??P_ 01 \
+
+                }
+            }
+        } else {
+            if { $async_set == 1 } {
+                if { $async_reset == 1 } {
+                    yosys log "Legalize DFFs for async set"
+                    yosys log "Legalize DFFs for async reset"
+                    yosys dfflegalize \
+                        -cell \$_DFF_P_ 01 \
+                        -cell \$_DFF_PN?_ 01 \
+                        -cell \$_DFFSR_PNN_ 01 \
+
+                } else {
+                    yosys log "Legalize DFFs for async set"
+                    yosys dfflegalize \
+                        -cell \$_DFF_P_ 01 \
+                        -cell \$_DFF_PN1_ 01 \
+
+                }
+            } else {
+                if { $async_reset == 1 } {
+                    yosys log "Legalize DFFs for async reset"
+                    yosys dfflegalize \
+                        -cell \$_DFF_P_ 01 \
+                        -cell \$_DFF_PN0_ 01 \
+
+                } else {
+                    # Choose to legalize to async resets even though they
+                    # won't tech map.  Goal is to get the user to fix
+                    # their code and put in synchronous resets
+                    yosys log "Legalize DFFs for synchronous reset only"
+                    yosys dfflegalize \
+                        -cell \$_DFF_P_ 01 \
+                        -cell \$_DFF_P??_ 01 \
+
+                }
+            }
+        }
+    }
+}
```

Comment:
this can probably be simplified into a set of FFs based on the attributes and then just call the dfflegalize once with `{*}$args` type of thing. We can talk about this

Reply:
Trying to build the arguments out of a list produces this error:

`dfflegalize -cell {$_DFF_P_} 01 -cell {$_DFF_PN?_} 01 -cell {$_DFFE_PP_} 01 -cell {$_DFFE_PN?P_} 01 -cell {$_DFFSR_PNN_} 01 -cell {$_DFFSR_PNN_} 01`

There are some issues with needing to escape character the leading \ and $ confounding this; second opinions welcome.  Just eliminating log statements and changing the nested if/else to if ... elseif ... else helps readability to my eye.

Reply:
Ah, perhaps I don't understand the significance of the {*}$args syntax.  Stay tuned

Reply:
Yes; [this stackoverflow thread](https://stackoverflow.com/questions/37926222/what-is-the-significance-of-when-used-with-args-like-args-in-tcl) was helpful.